optimum/habana/transformers/models/mpt/modeling_mpt.py

```diff
@@ -30,144 +37,288 @@
+            prev[:, :, :inp_seq_len, :].copy_(cur)
+            return orig_cur
+        assert cur.shape[2] == 1, f"Cannot update kv-cache. Unsupported shapes. prev:{prev.shape} cur:{cur.shape}"
+        if idx is not None:
+            prev.index_copy_(dim, idx - 1, cur)
+            return prev
+        else:
+            return torch.cat((prev, cur), dim=dim)
+
+    def get_shape(self):
+        if self.cache is None:
+            return None
+        return self.cache.shape
+
+    def forward(self, cur, dim, idx):
+        return self.update(self.cache, cur, dim, idx, self.inp_seq_len)
+
+
+class GaudiMptAttention(MptAttention):
+    def __init__(self, config: MptConfig):
+        super().__init__(config)
+        self.config = config
+        self.n_heads = self.config.n_heads
+
+        self.matmul_qk = Matmul()
+        self.matmul_av = Matmul()
+        self.k_cache = KVCache()
+        self.v_cache = KVCache()
+        self.inp_seq_len = -1
+
+    def allocate_kv_cache(self, batch_size, seq_len, inp_seq_len):
+        kv_shape = (batch_size, self.n_heads, seq_len, self.head_dim)
+        device = self.Wqkv.weight.device
+        dtype = self.Wqkv.weight.dtype
+        self.k_cache.allocate(inp_seq_len, dtype, device, kv_shape)
+        self.v_cache.allocate(inp_seq_len, dtype, device, kv_shape)
+
+    def reorder(self, tensor, beam_idx, dim_a, dim_b):
+        updated = tensor.index_select(0, beam_idx)
+        tensor.copy_(updated)
+
+    def reorder_kv_cache(self, beam_idx: torch.LongTensor):
+        if self.k_cache.cache is None:
+            return (None, None)
+
+        head_dim = self.k_cache.cache.size(-1)
+        seq_length = self.k_cache.cache.size(-2)
+        self.reorder(self.k_cache.cache, beam_idx, seq_length, head_dim)
+        self.reorder(self.v_cache.cache, beam_idx, seq_length, head_dim)
+        return (self.k_cache.cache.shape, self.v_cache.cache.shape)
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        position_bias: torch.Tensor,
+        past_key_value: Optional[Tuple[torch.Tensor]] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        use_cache: bool = False,
+        token_idx: Optional[torch.Tensor] = None,
+        reuse_cache: Optional[bool] = False,
+        cache_idx: Optional[int] = None,
+    ):
+        """
+        Copied from MptAttention.forward: https://github.com/huggingface/transformers/blob/v4.32.0/src/transformers/models/mpt/modeling_mpt.py
+        The only differences are:
+        - add new args token_idx
+        - optimize KV cache
+        - add new args use_cache
+        - add new args reuse_cache
+        - add new args cache_idx
+        """
+        batch_size, seq_length = hidden_states.shape[:2]
+        mixed_qkv = self.Wqkv(hidden_states)
+        bs, seq_len, three_times_hidden_size = mixed_qkv.shape
+        mixed_qkv = mixed_qkv.view(bs, seq_len, self.n_heads * 3, self.head_dim)
+        mixed_qkv = mixed_qkv.transpose(1, 2)
```

Comment:
```suggestion
        mixed_qkv = mixed_qkv.view(bs, seq_len, self.n_heads * 3, self.head_dim).transpose(1, 2)
```

matter of personal taste. Ignore if you disagree!